modules/Channel.md

```diff
@@ -0,0 +1,112 @@
+## Channel Module
+
+A channel is a FIFO data structure used to communicate between different Chapel
+tasks. You can learn more about how go-channels work by following this
+[link](https://gobyexample.com/channels)
+Since the operations on the channel are blocking by default, it enables to send
+values from one task to another in a synchronized manner.
+Here is a simple example in Go.
+
+```
+func main() {
+	// make a channel for string values
+	messages := make(chan string)
+
+	// new task to send a value to the channel
+	// it will wait until other task receives
+	go func() { messages <- "ping" }()
+
+	// receive the sent value
+	msg := <-messages
+	fmt.Println(msg)
+}
+```
+
+The module will have a generic `Channel` class (or record) to store the channel
+buffer and support operations on them.
+
+* The main functions that the module should have
+
+	* `init(type eltType, size : int)` : Specify the buffer data type and size
+	for creating a Channel.
+
+	* `send(value : eltType)` : This function waits for a space in the buffer
+	and	sends the `value` to the channel.
+
+	* `recv(out value : eltType) : bool` : This function waits until there is
+	something in the buffer and stores the received entity in `value`. It also
+	returns the status i.e., the operation is executed successfully or not.
+
+	* `close()` : This function is used to close the channel such that no more
+	values will be sent to it.
+
+* `iter these()` : An iterator to receive the values sent to the channel until
+it is closed.
+
+* `select` statements : This statement lets you wait for multiple channel
+operations and executes if anyone of them is ready for communication. `select`
+statements in Go are specifically designed to work with channels.
+Some key points are:
+
+	* `select` statements are commonly used in Go code when working with channels.
+	* The operations are atomic, i.e., checking a channel and performing `recv`
+		or `send` on it is considered as one operation.
+	* The cases are arranged in a random order to prevent starvation.
+
+For e.g.
+```
+// Select syntax in Go to wait on any one of the two channels for receiving
+// If every channel is blocked, the default case is executed.
+select {
+	case msg1 := <-chan1 :
+		fmt.Println("Received", msg1)
+	case msg2 := <-chan2 :
+		fmt.Println("Received", msg2)
+	default :
+		fmt.Println("None")
+}
+```
+
+```
+// Select statements can be used to implement try-send or try-recv
+// default case is executed when none of the cases are ready yet.
+select {
+	case val := <-chan1 :
+		fmt.Println("Received", val);
+	default :
+		fmt.Println("Channel not ready");
+}
+```
+
+### Some important questions related to the module-
+
+* Should this module be standard or package ?
+
+* What should be the name of the module ? Some names in other languages
+	* `chan` in Golang
+	* `crossbeam` in Rust
```

Comment:
I personally feel `crossbeam` might distinguish it from being confused with IO Channels.

Reply:
Would we want to unify IO channels with these channels in some way?  (probably a question for @mppf).  If not, how would we describe their differences.

Whatever we call it, we should use an initial-capital name.  If we used a term more like Channels, I'd spell out the word rather than using `chan` (which feels like an unnecessary abbreviation to me, and maybe an unfortunate one in the era of 4chan and the like).  If we decided to make this a standard module, we might think about whether it should be merged with some other module (like I/O, e.g.)

I'd start by focusing more on the semantics, behavior, implementation, and type and procedure names than the module name in any case, since it's easy enough to rename modules or move code between modules, particularly when they're just starting out.

Reply:
The I/O channels and these Go channels seem pretty different, on the surface, in terms of several key properties:
 * Go-style channels will be both readable and writable but I/O channels are either reading or writing but not both
 * The normal use case for the Go-style channel is to moved typed data (e.g. ints or strings or whatever) but the I/O channel moves bytes (i.e. everything is serialized). In this way they would be more similar to ItemReader / ItemWriter that is in the library today.
  * Go-style channels have some important concurrency / atomicity properties that might be difficult to retrofit on to the I/O channels

While the names are related (and I did imagine we might use a `channel` for cross-locale communication, say), to me the part about moving typed data makes them different enough to justify a different name. Of course, we could try to make all `channel`s have an `eltType` and then try to expand `channel` to have the properties needed for Go-style channels but - Why? Is it just because we couldn't come up with a better name? It seems to me that these two types have pretty different use cases.

So it's my position that these should be different types and we need to find a solution to the name collision. I would personally be fine with `crossbeam` or `crossbar` especially if those terms are used to mean this in other projects. We can update the documentation for `channel` to point to the new type in case there is anybody searching and ending up in the wrong place.

Reply:
I also think unifying them with the IO module would create more confusion, as they don't perform any IO. 
I am fine with  `CrossBeam` or `CrossBar` (given that module names would start with a capital letter).

Reply:
The `cross` names don't work for me because they imply structural complexity rather than just something that allows moving a substance from once place to another. (And because a "crossbar" is an actual thing: in hardware, a kind of multiplexed switch.) I'm not fond of `channel` (however capitalized) for the module or type or even concept here, because it implies a similarity with IO channels which isn't present. But it's the right sort of idea, so some similar word might work. `Pipe` is close, because these do have certain similarities with POSIX pipes though obviously they're also not identical. Do any of `conduit` (probably out because of potential confusion with the GASNet use of that term), `hose`, `duct`, or `tube` work for folks?

Reply:
It occurs to me that we might be able to rename the I/O `channel` to `reader` / `writer` - I bring this up in issue #18112. Interested in feedback.

Regarding `pipe` / `conduit` / `hose` / `duct` / `tube` - I find `conduit` relatively appealing of these.

Reply:
Yeah, `Pipe`  seems like an interesting choice to me, but if we could rename the I/O `channel`, that would be best.

Reply:
Since the terms `Pipe`([link](https://gobyexample.com/channels)) and `Conduit`([link](https://tour.golang.org/concurrency/2)) are used for describing channels in Go official docs, I think both would work. But if `Pipe` would create confusion with POSIX pipes, I think `Conduit` would be more suitable in that case.

Reply:
It occurred to me last night that `channel` is what the conveyances are called in [Communicating Sequential Processes](https://en.wikipedia.org/wiki/Communicating_sequential_processes), so I've changed my mind and that's now my favorite, assuming our I/O `channel` can be renamed. `Conduit` has the conflict I noted, in that GASNet (our most-portable communication configuration) uses that for the embodiments of its abstract communication model.